modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

```diff
@@ -11,16 +11,366 @@ redirect_from:
+  //sampleEnd
+  println("list= $list")
+  println("isLeft= $isLeft")
+}
+```
+
+```kotlin:ank:playground
+import arrow.core.Either
+import arrow.core.Nel
+import arrow.core.ValidatedNel
+import arrow.core.extensions.nonemptylist.semigroup.semigroup
+import arrow.core.extensions.validated.applicative.applicative
+import arrow.core.fix
+import arrow.core.invalidNel
+import arrow.core.k
+import arrow.core.validNel
+
+fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+  if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+  else Either.Left(NumberFormatException("$s is not a valid integer."))
+
+fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+  if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+  else NumberFormatException("$s is not a valid integer.").invalidNel()
+
+fun main() {
+  //sampleStart
+  val validatedList = listOf("1", "22w", "3", "33s").k()
+    .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
+  //sampleEnd
+  println(validatedList)
+}
+```
+
+Notice that in the `Either` case, should any string fail to parse the entire `traversal` is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option` as the effect). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of traversal is closely tied with the `Applicative` behavior of the data type.
+
+Going back to our `Promise` example, we can get an `Applicative` instance for `IO`, by converting `Promise` to `IO` that runs each `Promise` concurrently. Then when we traverse a `List<A>` with an `(A) -> Promise<ForIO, B>`, we can imagine the traversal as a scatter-gather. Each `A` creates a concurrent computation that will produce a `B` (the scatter), and as the `Promise`s complete they will be gathered back into a `List`.
+
+{TODO: finish example with Promise}
+
+```kotlin:ank:playground
+import arrow.core.Either
+import arrow.core.Option
+import arrow.fx.Promise
+
+interface Profile
+interface User
+
+fun parseInt(s: String): Option<Int> = TODO()
+data class SecurityError(val message: String)
+data class Credentials(val email: String, val password: String)
+
+fun validateLogin(c: Credentials): Either<SecurityError, Unit> = TODO()
+
+fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
+
+data class Admin(val name: String, val crd: Credentials): User
+
+fun ListK<User>.processLogin()
+```
+
+Evidently, `Traverse` is not limited to `List` or `Nel`, it provides an abstraction over 'things that can be traversed over', like a Binary tree, hence the name `Traverse`.
+
+#### Playing with `Reader`
+
+Another interesting effect we can use is Reader. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.
```

Comment:
A bit more context. Reader is an `injector for a parameter` or something similar.